PacketDotNet/UdpPacket.cs

```diff
@@ -76,17 +94,14 @@ public UdpPacket(ByteArraySegment byteArraySegment)
 
             PayloadPacketOrData = new LazySlim<PacketOrByteArraySegment>(() =>
             {
-                const int wakeOnLanPort0 = 0;
-                const int wakeOnLanPort7 = 7;
-                const int wakeOnLanPort9 = 9;
-                
                 var result = new PacketOrByteArraySegment();
                 var destinationPort = DestinationPort;
                 var sourcePort = SourcePort;
                 var payload = Header.NextSegment();
 
                 // If this packet is going to port 0, 7 or 9, then it might be a WakeOnLan packet.
-                if (destinationPort == wakeOnLanPort0 || destinationPort == wakeOnLanPort7 || destinationPort == wakeOnLanPort9)
+                if (PayloadTypeDestinationPortMappingList.ContainsKey(destinationPort) &&
+                    PayloadTypeDestinationPortMappingList[destinationPort] == typeof(WakeOnLanPacket))
```

Comment:
Thanks. See some more comments 🙂

I don't understand the need for this, these ports are always WOL packets. The other port is always L2TP.
I would keep this code as-is or maybe change this bit to modern C# as to tidy it up a bit:
```if (destinationPort is wakeOnLanPort0 or wakeOnLanPort7 or wakeOnLanPort9)```

Furthermore, as RTP doesn't use fixed ports. there seems to be a way to somewhat reliably detect if it's a possible RTP packet (and keep payload in other cases), see: https://stackoverflow.com/questions/2916177/how-do-i-determine-if-a-packet-is-rtp-rtcp 
Alternatively, we can perhaps move the ability to add custom RTP ports to ```RtpFields``` and expose these ports through an https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.iset-1?view=net-5.0.

Reply:
Hi,
Regarding the port mapping, see https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers, the 0, 7 and 9 are actually used for WOL packets but only unofficial, else this is reserved for Echo and could be reused. I saw this post https://github.com/chmorgan/packetnet/issues/103.

Now for RTP/RTCP, the only really option we have would be to decode the SDP bodies of the SIP messages, but this is really complex and might be impossible because of IPsec and encryption. So custom RTP ports (like the wirkeshark "Decode As") seems to be the only option.

Reply:
The idea behind the Destination and Source lists of ports at Udp class level (and not in RtpFields) was here to avoid any conflict across the different payload type. Each port is then unique and with a single type of packet.

Reply:
> 
> 
> Hi,
> Regarding the port mapping, see https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers, the 0, 7 and 9 are actually used for WOL packets but only unofficial, else this is reserved for Echo and could be reused. I saw this post #103.
> 

In theory, yes. But in practise it is almost always sent as a UDP datagram to port 0, 7 (Echo) or 9 (Discard) 🙂


> Now for RTP/RTCP, the only really option we have would be to decode the SDP bodies of the SIP messages, but this is really complex and might be impossible because of IPsec and encryption. So custom RTP ports (like the wirkeshark "Decode As") seems to be the only option.

RTP headers should always exist, https://datatracker.ietf.org/doc/html/rfc3550#section-5.1 I think it might be possible to get a pretty accurate determination for RTP packets based on the data. Similar to DHCPv4 is being checked by UDP.

That said, a "decode as" option isn't a bad idea for Packet.Net either. That's already somewhat possible, as you can directly set the PayloadPacket. Perhaps we can add a something like a collection of Func<UdpPacket, PacketOrByteArraySegment> with custom parsers.

Reply:
Except based on port, there is no safe way based only on the bytes to confirm if a UDPP payload is RTP or not. And we don't want to try the Rtp decoding on all Udp packets neither. The RTP doesn't include any checksum nor any length for its own payload, so it makes easy to interpret / decode any random UDP payload as a valid RTP even if it's not RTP.

Now the idea of a function is good, and can be offered additionally. Would be possible to do this in another PR?

Reply:
"As it turns out, checking the bytes in the RTP header and checking for basically the version and payload type are enough to determine if it's an RTP packet. At least so far I haven't found any other packets on the network that have the same first few bits."

It's pretty similar to how DHCPv4 and WOL packets are checked, this is also done by comparing to expected data. It would probably mean you'd need to add the possible payload types. Wireshark has a list of them. This would be more in-line with the rest of the library as it _tries_ to parse packets based on the data that's available. With the other payload packets it's not a guaranteed result either, but one can always manually set the payload to another packet type.

As for custom decoders: let me think on that for a bit, I don't think payload port mappings like in this PR is the best solution as you're completely bound to ports then. Perhaps something like:
Dictionary<Type, PayloadDecoder> PayloadDecoders
delegate PacketOrByteArraySegment PayloadDecoder(Packet parentPacket, ByteArraySegment payload)

Then all existing decoders can be added in there. If someone then needs to change the decoder then it's possible to replace it, or add a custom decoder for custom types. But that would be something for another PR.

Reply:
Do I still have to change something here or this will be done in another PR and this one can be approved?
I would like to continue with RTCP.
Thanks.

Reply:
> 
> 
> Do I still have to change something here

Yes, check the above reply 🙂

Reply:
sorry, can you be more specific, because I don't get it. There is no certain way to check that a byte array is RTP or not (can be RTCP or something else) except based on the ports used.

Reply:
If you check the post I've linked to it says it's possible to check RTP based on the header's version and payload type. You can possibly use more such as the size of the payload as it should always be 32-bit aligned going by the RFC. Also CC + length of length of the CSRC section. There's no certain way, but that is also the case for the other packets such as DHCP and WOL.

Reply:
ok, then I think we are blocked here, because I disagree. Decoding a UDP payload only based on this criteria will only bring failures.
I will try then the other approach (with a fallback function) or look for another library.
Thanks

Reply:
@jgaulon Just on port will also bring failures 🙂

Reply:
If it wasn't clear, I think this is a great addition 👍 But this project is always trying to guesstimate the payload based on data, and it would be strange to suddenly not do that here. It seems pretty trivial to check the RTP header to have few false positives. This is also done for WOL and DHCPv4 in the UDP packet class you've modified. Failures will always occur (it also happens in Wireshark), which is why it's also possible to read the raw packet and set the payload type manually. 

Custom decoders would be a useful addition in general so we should certainly look into that for another PR.

Reply:
Thanks.